kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -279,6 +282,385 @@ impl<'tcx> GotocCtx<'tcx> {
+                    let op_expr = self.codegen_operand(o);
+                    let op_width = op_expr.typ().sizeof_in_bits(&self.symbol_table);
+                    if op_width == 0 {
+                        None
+                    } else {
+                        Some((
+                            InternedString::from(f.name.to_string()),
+                            op_expr.transmute_to(c.typ(), &self.symbol_table),
+                        ))
+                    }
+                })
+                .collect(),
+            &self.symbol_table,
+        )
+    }
+
+    fn codegen_rvalue_enum_direct(
+        &mut self,
+        variant_index: &VariantIdx,
+        operands: &[Operand<'tcx>],
+        overall_t: Type,
+        adt: &AdtDef<'_>,
+        variants: &IndexVec<VariantIdx, LayoutS<VariantIdx>>,
+    ) -> Expr {
+        let fields = overall_t.lookup_components(&self.symbol_table).unwrap().clone();
+        assert_eq!(fields.len(), 2, "TagEncoding::Direct encountered for enum with empty variants");
+        assert_eq!(
+            fields[0].name().to_string(),
+            "case",
+            "Unexpected field in enum/generator. Please report your failing case at https://github.com/model-checking/kani/issues/1465"
+        );
+        let case_value = Expr::int_constant(variant_index.index(), fields[0].typ());
+        assert_eq!(
+            fields[1].name().to_string(),
+            "cases",
+            "Unexpected field in enum/generator. Please report your failing case at https://github.com/model-checking/kani/issues/1465"
+        );
+        assert!(matches!(variants[*variant_index].variants, Variants::Single { .. }));
+        let variant = &adt.variants()[*variant_index];
+        if variant.fields.is_empty() {
+            Expr::struct_expr_with_nondet_fields(
+                overall_t,
+                btree_string_map![("case", case_value)],
+                &self.symbol_table,
+            )
+        } else {
+            let target_component = fields[1]
+                .typ()
+                .lookup_field(variant.name.to_string(), &self.symbol_table)
+                .unwrap()
+                .clone();
+            let cases_value = Expr::union_expr(
+                fields[1].typ(),
+                target_component.name(),
+                Expr::struct_expr_from_values(
+                    target_component.typ(),
+                    variants[*variant_index]
+                        .fields
+                        .index_by_increasing_offset()
+                        .map(|idx| self.codegen_operand(&operands[idx]))
+                        .collect(),
+                    &self.symbol_table,
+                ),
+                &self.symbol_table,
+            );
+            Expr::struct_expr_from_values(
+                overall_t,
+                vec![case_value, cases_value],
+                &self.symbol_table,
+            )
+        }
+    }
+
+    fn codegen_niche_struct_from_operands_rec(
```

Comment:
I don't think we should duplicate the niche calculation. I was wondering if we could leverage the `SetDiscriminant` logic instead. If you move that logic to its own function, you could reuse it here.

Reply:
Done, will produce a separate PR for this change.